java/java.lsp.server/vscode/src/extension.ts

```diff
@@ -163,14 +163,45 @@ export function activate(context: ExtensionContext) {
     }));
 }
 
+/**
+ * Pending maintenance (install) task, activations should be chained after it.
+ */
+let maintenance : Promise<void> | null;
+
+/**
+ * Pending activation flag. Will be cleared when the process produces some message or fails.
+ */
+let activationPending : boolean = false;
+
 function activateWithJDK(specifiedJDK: string | null, context: ExtensionContext, log : vscode.OutputChannel, notifyKill: boolean): void {
-    if (nbProcess) {
+    const a : Promise<void> | null = maintenance;
+    if (activationPending) {
+        // do not activate more than once in parallel.
+        return;
+    }
+    activationPending = true;
+    if (a != null) {
+        a.then(() => doActivateWithJDK(specifiedJDK, context, log, notifyKill));
+    } else {
+        doActivateWithJDK(specifiedJDK, context, log, notifyKill);
+    }
+}
+
+function killNbProcess(notifyKill : boolean, specProcess?: ChildProcess) {
+    let p = nbProcess;
+    if (p && (!specProcess || specProcess == p)) {
         if (notifyKill) {
             vscode.window.setStatusBarMessage("Restarting Apache NetBeans Language Server.", 2000);
         }
-        nbProcess.kill();
+        p.kill();
+        // PENDING: possible race
```

Comment:
What do you mean by possible race, @sdedic?

Reply:
leftover; Promise reordering no longer happens, and the nodejs runtime schedules all callbacks in a single thread, right ?

Reply:
Everything in one extension runs in a single process.